napari/_vispy/vendored/volume.py

```diff
@@ -582,13 +583,24 @@
                     color = calculateColor(color, iloc, dstep);
                     gl_FragColor = applyColormap(color.r);
                     iter = nsteps;
+                    
+                    //Set depth value
+                    float clip_distance = length(farpos - nearpos);
+                    vec3 surface_point = iloc * u_shape;
+                    float surface_dist = length(surface_point - nearpos);
+                    float depth = surface_dist / clip_distance;
+                    gl_FragDepth = depth;
+                    surface_found = 1;
                     break;
                 }
                 iloc += step * 0.1;
             }
         }
         """,
     after_loop="""
+        if (surface_found == 0) {
+            discard;
+        }
```

Comment:
I have something similar in the plane rendering PR - we should make sure to coordinate here!

Out of interest, did you get weird blue/green background without this or just black?

Reply:
cc @brisvag for clipping planes where we added discard if no data sampled for a fragment

Reply:
Yup, this looks exactly the same! Once those are merged, I'm sure this can be updated easily.

Reply:
We can consider adding a flag that all of the modes can use to discard when the fragment is empty. Perhaps it can go after the "after_loop" part of the shader. One thing to consider is that it is advantageous to discard as soon as you know you can to prevent unnecessary operations. that being said, in this case, the difference is probably (i think?) small.

Reply:
See [this comment](https://github.com/napari/napari/pull/3140#discussion_r690202697) for context.